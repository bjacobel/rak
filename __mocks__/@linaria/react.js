export const styled = jest.fn(strings => () => strings);
