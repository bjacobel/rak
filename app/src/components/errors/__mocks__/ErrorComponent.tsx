export default () => (
  <>
    <h3 className="errorHeader">Something went wrong.</h3>
    <pre>In the real component, the stack would be rendered here.</pre>
  </>
);
