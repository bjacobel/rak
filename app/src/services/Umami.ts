/* eslint-disable class-methods-use-this */

import { TRACK_ANALYTICS, UMAMI_ID } from '../constants';

declare global {
  interface Window {
    umami: {
      track: (propsFn: (existingProps: Record<string, string>) => Record<string, string>) => void;
    };
    scriptLoaded: boolean;
  }
}

window.scriptLoaded = false;

export default class Umami {
  constructor() {
    if (!window.scriptLoaded) {
      const script = document.createElement('script');
      script.defer = true;
      script.src = 'https://cloud.umami.is/script.js';
      script.setAttribute('data-website-id', UMAMI_ID);
      script.setAttribute('data-auto-track', 'false');
      document.body.appendChild(script);

      window.scriptLoaded = true;
    }
  }

  track(propsFn: (existingProps: Record<string, string>) => Record<string, string>) {
    if (window.umami && TRACK_ANALYTICS) {
      window.umami.track(propsFn);
    }
  }
}
