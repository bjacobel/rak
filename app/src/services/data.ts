import { DATA } from '../constants';

export type Data = typeof DATA;

// eslint-disable-next-line import/prefer-default-export
export const getData = () => Promise.resolve(DATA);
