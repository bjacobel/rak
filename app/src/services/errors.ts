import { BrowserMicroSentryClient } from '@micro-sentry/browser';
import { BreadcrumbsPlugin } from '@micro-sentry/breadcrumbs-plugin';

import { LOG_ERRORS, RAVEN_ENDPT, RELEASE } from '../constants';

let client: BrowserMicroSentryClient;

export const setup = () => {
  if (LOG_ERRORS) {
    if (!client) {
      client = new BrowserMicroSentryClient({
        dsn: RAVEN_ENDPT,
        release: RELEASE,
        plugins: [BreadcrumbsPlugin],
      });
    }
  }
};

export default (ex: Error, context?: unknown) => {
  if (LOG_ERRORS) {
    setup(); // memoized, it is fine to call this on every log

    if (context) {
      client.setExtra('context', context as string);
    }

    client.report(ex);
  }

  return window.console && console.error && console.error(ex);
};
