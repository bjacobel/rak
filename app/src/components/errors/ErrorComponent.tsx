import { styled } from '@linaria/react';

import { SHOW_STACK } from '../../constants';
import { robotoMono } from '../../stylesheets/shared.css';

const MonoHeader = styled.h3`
  ${robotoMono}
`;

const ErrorComponent = ({ error }: { error?: Error }) => (
  <>
    <MonoHeader>Something went wrong.</MonoHeader>
    {SHOW_STACK && error && error.stack && <pre data-testid="stack">{error.stack}</pre>}
  </>
);

export default ErrorComponent;
