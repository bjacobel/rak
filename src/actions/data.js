import { loadingStarted, loadingEnded } from './loading';
import { getData } from '../services/data';
import logToRaven from '../services/errors';

export const GET_DATA_FAILED = 'GET_DATA_FAILED';
export const GET_DATA_SUCCEEDED = 'GET_DATA_SUCCEEDED';

export const getDataSucceeded = data => ({
  type: GET_DATA_SUCCEEDED,
  payload: { data },
});

export const getDataFailed = err => {
  logToRaven(err);
  return { type: GET_DATA_FAILED, payload: { errors: [err] } };
};

export const getDataAsync = () => dispatch => {
  dispatch(loadingStarted());

  return getData()
    .then(data => {
      dispatch(loadingEnded());
      dispatch(getDataSucceeded(data));
    })
    .catch(err => {
      dispatch(loadingEnded());
      dispatch(getDataFailed(err));
    });
};
