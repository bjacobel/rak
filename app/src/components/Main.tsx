import { useQuery } from '@tanstack/react-query';
import { styled } from '@linaria/react';

import { getData, Data } from '../services/data';
import ErrorComponent from './errors/ErrorComponent';
import { lightgrey } from '../stylesheets/colors.css';
import { center, openSans } from '../stylesheets/shared.css';
import StyledLink from './StyledLink';

const DataHeader = styled.h3`
  ${center}
  ${openSans}
  ${lightgrey}
  font-style: italic;
`;

const Logo = styled.div`
  height: 100px;
  background-image: url('../assets/images/logo.svg');
  background-repeat: no-repeat;
  background-position: center;
`;

export default () => {
  const { isLoading, data, error } = useQuery<Data, Error>({ queryKey: ['data'], queryFn: getData });

  if (error) {
    return <ErrorComponent error={error} />;
  }

  if (isLoading) {
    return <span>loading...</span>;
  }

  return (
    <>
      <Logo role="banner" />
      <DataHeader>{data!.text || ''}</DataHeader>
      <StyledLink to="/child/foo">Routing demo</StyledLink>
      <StyledLink to="/asdf">404 demo</StyledLink>
      <StyledLink to="/error">Error</StyledLink>
    </>
  );
};
