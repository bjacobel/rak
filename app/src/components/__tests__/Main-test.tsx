import { ReactNode } from 'react';
import { Switch, Route } from 'wouter';

import { getData } from '../../services/data';
import { DATA } from '../../constants';
import Main from '../Main';
import { render, waitFor } from '../../testing/rtl';
import ErrorComponent from '../errors/ErrorComponent';

jest.mock('../../services/data');

jest.mock('../errors/ErrorComponent', () => jest.fn(() => <span>ErrorComponent</span>));

describe('main component', () => {
  let main: ReactNode;

  beforeEach(() => {
    jest.mocked(getData).mockImplementation(() => Promise.resolve(DATA));
    main = (
      <Switch>
        <Route component={Main} path="/" />
      </Switch>
    );
  });

  it('matches snapshot', async () => {
    const { asFragment, getByRole } = render(main);

    await waitFor(() => expect(getByRole('banner')).toBeInTheDocument());

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders an h3 with data if it didn't hit an error", async () => {
    const { getByText } = render(main);

    await waitFor(() => expect(getByText(DATA.text)).toBeInTheDocument());
  });

  it('renders error if data-fetching errors', async () => {
    const error = new Error('mock error');
    jest.mocked(getData).mockImplementationOnce(() => {
      throw error;
    });

    const { asFragment } = render(main);

    await waitFor(() => expect(ErrorComponent).toHaveBeenCalled());

    expect(asFragment()).toMatchSnapshot();
  });
});
