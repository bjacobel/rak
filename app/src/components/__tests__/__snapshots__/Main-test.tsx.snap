// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main component matches snapshot 1`] = `
<DocumentFragment>
  <div
    role="banner"
  />
  <h3>
    an opinionated skeleton to quickly set up a new React app
  </h3>
  <a
    href="/child/foo"
  >
    Routing demo
  </a>
  <a
    href="/asdf"
  >
    404 demo
  </a>
  <a
    href="/error"
  >
    Error
  </a>
</DocumentFragment>
`;

exports[`main component renders error if data-fetching errors 1`] = `
<DocumentFragment>
  <span>
    ErrorComponent
  </span>
</DocumentFragment>
`;
